specification/sections/05-events.md

```diff
@@ -147,22 +154,25 @@ See [provider initialization](./02-providers.md#24-initialization), [setting a p
 
 ### Event handlers and context reconciliation
 
-Providers built to conform to the static context paradigm feature an additional `PROVIDER_CONTEXT_CHANGED` event, which is used to signal that the global context has been changed, and flags should be re-evaluated.
+Providers built to conform to the static context paradigm feature two additional events: `PROVIDER_CONTEXT_CHANGE_PENDING` and `PROVIDER_CONTEXT_CHANGED`.
+While the provider is reconciling it's internal state (the `on context changed` function is running and not yet terminated) the SDK emits `PROVIDER_CONTEXT_CHANGE_PENDING` and transitions the provider into state `CONTEXT_PENDING`.
+This can be particularly useful for displaying loading indicators while the [evaluation context](./03-evaluation-context.md) is being reconciled.
+
+If the `on context changed` function terminates normally, the SDK emits (`PROVIDER_CONTEXT_CHANGED`) and transitions the provider into the `READY` state, otherwise it emits `PROVIDER_ERROR` and transitions the provider into `ERROR` state.
```

Comment:
How do we handle cases where `setEvaluationContext()` is called a second time by the application before the first call has not finished yet? Should the first `onContextChange()` call terminate normally or abnormally, and what are the implications with the Events emitted and the State changes?

Reply:
That's a great question.

My first thoughts are that we shouldn't terminate abnormally, and should emit events as normal. We may want to consider adding requirements on providers to make sure the function is reentrant, so at least there's no accidental errors if this happens.

We could also guarantee this doesn't happen by queueing up context-changes, but I think that's putting quite a burden on the SDK for a bit of an edge case. Seeing that `setContext` is an async operation in most implementations, I think authors inherently expect that multiple calls might not finish in the order they are started, and so should do what they can to avoid such race conditions.

A middle ground might be to never emit `PROVIDER_CONTEXT_CHANGED` or change the state from `CONTEXT_PENDING` while another reconciliation is in progress, which seems reasonable, so at least if the function is called twice, we wouldn't pop in and out of the pending state.

Reply:
Is this really an edge-case? I would imagine apps call setContext from event handlers (say a dark-mode toggle) which would quickly lead to overlapping reconciliation cycles, and the spec should be clear on how to handle these.

Also, from an experimentation perspective, it's adamant that the state remain in RECONCILING until the last initiated cycle has ended. If the state transitions to READY before the last cycle has ended, flags will return stale values (for example values for another user). If the state transitions to ERROR before the last cycle has ended, flags will return default values. Both of these behaviours will taint the user experience and the experiment.

Reply:
> Is this really an edge-case? I would imagine apps call setContext from event handlers (say a dark-mode toggle) which would quickly lead to overlapping reconciliation cycles, and the spec should be clear on how to handle these.
> 
> Also, from an experimentation perspective, it's adamant that the state remain in RECONCILING until the last initiated cycle has ended. If the state transitions to READY before the last cycle has ended, flags will return stale values (for example values for another user). If the state transitions to ERROR before the last cycle has ended, flags will return default values. Both of these behaviours will taint the user experience and the experiment.

@andreas-karlsson you make some compelling points.

We could keep things vague and just say SDKs must "handle" this in some way... but I'm thinking it's better to prescribe a pattern to handle it.

Do you think if we spec'd something like what I described [here](https://github.com/open-feature/spec/pull/241#discussion_r1479925311) (specifically the quoted section below) it would be sufficient?

> A middle ground might be to never emit PROVIDER_CONTEXT_CHANGED or change the state from CONTEXT_PENDING while another reconciliation is in progress, which seems reasonable, so at least if the function is called twice, we wouldn't pop in and out of the pending state.

cc @fabriziodemaria

Reply:
Thanks, that would be great.

Reply:
@andreas-karlsson @fabriziodemaria I've pushed a commit to address this specific issue: https://github.com/open-feature/spec/pull/241/commits/ad3b33768f9de746b9c766164b186b62d7dcaddc